From b40e549127e00fbd6426fac40a7033a893517b9e Mon Sep 17 00:00:00 2001
From: Liu Bolun <bolunx.liu@intel.com>
Date: Tue, 11 Sep 2012 11:29:48 +0800
Subject: [PATCH] [PORT FROM MAIN] Adding free heap when start encode fail.

we should hanlde error occur in initialization that calling
MediaSource::stop()when mState equals to LOADED_TO_IDLE,
it means OMXComponet is initializing only in initialization stage.
In addition, we not only stop mediasource, but also free buffer for
OMXCodec ports avoid memory leak.Modifying code style.

Change-Id: I1fe805721cc9ed65d80432faf034328be0725843
Category: aosp improvement
Domain: Video.Media-Framework
Origin: internal
Signed-off-by: Liu Bolun <bolunx.liu@intel.com>
Signed-off-by: Dan Liang <dan.liang@intel.com>

diff --git a/media/libstagefright/OMXCodec.cpp b/media/libstagefright/OMXCodec.cpp
index 269028a..0f58d48 100644
--- a/media/libstagefright/OMXCodec.cpp
+++ b/media/libstagefright/OMXCodec.cpp
@@ -3181,6 +3190,16 @@ void OMXCodec::onEvent(OMX_EVENTTYPE event, OMX_U32 data1, OMX_U32 data2) {
         {
             CODEC_LOGE("ERROR(0x%08lx, %ld)", data1, data2);
 
+            // we should hanlde error occur in initialization that calling MediaSource::stop()
+            // when mState equals to LOADED_TO_IDLE, it means OMXComponet is initializing
+            // only in initialization stage.
+            if(mState == LOADED_TO_IDLE)
+            {
+               freeBuffersOnPort(kPortIndexInput);
+               freeBuffersOnPort(kPortIndexOutput);
+               mSource->stop();
+            }
+
             setState(ERROR);
             break;
         }

